tools/deploy/helper.go

```diff
@@ -0,0 +1,442 @@
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/m3db/m3cluster/kv"
+	"github.com/m3db/m3cluster/services"
+	"github.com/m3db/m3cluster/services/placement"
+	"github.com/m3db/m3x/log"
+	"github.com/m3db/m3x/retry"
+	"github.com/m3db/m3x/sync"
+)
+
+var (
+	errNoDeploymentProgress = errors.New("no deployment progress")
+	errInvalidRevision      = errors.New("invalid revision")
+)
+
+// Mode is the deployment mode.
+type Mode int
+
+// A list of supported deployment modes.
+const (
+	DryRunMode Mode = iota
+	ForceMode
+)
+
+// Helper is a helper class handling deployments.
+type Helper interface {
+	// Deploy deploys a target revision to the instances in the placement.
+	Deploy(revision string, mode Mode) error
+}
+
+// TODO(xichen): disable deployment while another is ongoing.
+type helper struct {
+	logger             xlog.Logger
+	planner            planner
+	client             aggregatorClient
+	mgr                Manager
+	store              kv.Store
+	retrier            xretry.Retrier
+	foreverRetrier     xretry.Retrier
+	workers            xsync.WorkerPool
+	toPlacementIDFn    ToPlacementInstanceIDFn
+	toAPIEndpointFn    ToAPIEndpointFn
+	placementWatcher   services.StagedPlacementWatcher
+	settleBetweenSteps time.Duration
+}
+
+// NewHelper creates a new deployment helper.
+func NewHelper(opts HelperOptions) (Helper, error) {
+	client := newAggregatorClient(opts.HTTPClient())
+	planner := newPlanner(client, opts.PlannerOptions())
+	placementWatcherOpts := opts.StagedPlacementWatcherOptions()
+	placementWatcher := placement.NewStagedPlacementWatcher(placementWatcherOpts)
+	if err := placementWatcher.Watch(); err != nil {
+		return nil, err
+	}
+	retryOpts := opts.RetryOptions()
+	retrier := xretry.NewRetrier(retryOpts)
+	foreverRetrier := xretry.NewRetrier(retryOpts.SetForever(true))
+	return helper{
+		logger:             opts.InstrumentOptions().Logger(),
+		planner:            planner,
+		client:             client,
+		mgr:                opts.Manager(),
+		store:              opts.KVStore(),
+		retrier:            retrier,
+		foreverRetrier:     foreverRetrier,
+		workers:            opts.WorkerPool(),
+		toPlacementIDFn:    opts.ToPlacementInstanceIDFn(),
+		toAPIEndpointFn:    opts.ToAPIEndpointFn(),
+		placementWatcher:   placementWatcher,
+		settleBetweenSteps: opts.SettleDurationBetweenSteps(),
+	}, nil
+}
+
+func (h helper) Deploy(revision string, mode Mode) error {
+	if revision == "" {
+		return errInvalidRevision
+	}
+	all, err := h.allInstanceMetadatas()
+	if err != nil {
+		return fmt.Errorf("unable to get all instance metadatas: %v", err)
+	}
+	filtered := filterByRevision(all, revision)
+	plan, err := h.planner.GeneratePlan(filtered, all)
+	if err != nil {
+		return fmt.Errorf("unable to generate deployment plan: %v", err)
+	}
+
+	// If in dry run mode, log the generated deployment plan and return.
+	if mode == DryRunMode {
+		h.logger.Infof("Generated deployment plan: %+v", plan)
```

Comment:
I was thinking that it might be useful to the operator to know what the next steps are before they're run. For instance to know when a particular group was going to be deployed.

Here's how I'd probably run this tool:
(1) start a screen/tmux session
(2) do the dry run and ensure stuff makes sense
(3) do the actual run and `tee` to a log. 
(4) break out of the screen/tmux session
(5) follow along the `tee`d log to see what's going on. 

I can imagine not remembering the output of the dry run once the execution has started, or not `tee`-ing it. At which point I'd have to re-run the dry run mode to see what the executor is going to do next (would that even generate the same output btw?). Would be easier to just log the plan up-front before the execution. 

Feel free to disregard if you don't think that's a valid use-case.